tools/toolchains/gcc.py

```diff
@@ -82,6 +86,17 @@ def __init__(self, target,  notify=None, macros=None,
 
         self.flags["common"] += self.cpu
 
+        # Coverage is turned On if coverage_filter is not empty. This means all sources are compiled with coverage
+        # macro and application is linked with coverage flags.
+        # Only source files that match regex from coverage_filter list are compiled with coverage flags. Since turning
+        # On code coverage on all sources increases static data size it overlaps with stack allocation and causes linker
+        # error.
+        self.coverage_filter = coverage_filter
+        if self.coverage_filter:
+            self.flags["common"].append("-g")
+            self.flags["common"].append("-O0")
```

Comment:
Ok, thanks for clarifying. Considering that the mechanism for changing compile flags is usually delegated to the build profiles, I would say we'd probably want to stick the `--profile debug` option in this case when compiling with coverage.

It'd be possible before the compile check when using `--coverage-filter` to check to see if both `-g` and `-O0` are in the compiler flags and display an error to the user to use the `debug` profile.

So it'd be something like:

```
$ mbed test -t GCC_ARM --coverage-filter=...
[ERROR] Debug options must be set for code coverage to work properly. Please enable them by using the `--profile debug` option.

$ mbed test -t GCC_ARM --coverage-filter=... --profile debug
[Build completes]
```

I'd like to hear @theotherjimmy's opinion on this. I do think in the current implementation if someone were to execute `mbed test -t GCC_ARM --coverage-filter=... --profile debug` the `-g` and `-O0` would be added twice, so this should probably be resolved.